File: SinglyCircular_LL/SinglyCL3.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

void InsertFirst(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no;
    newn->next = NULL;

    if((*first == NULL)  && (*last == NULL))
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        newn->next = (*first);
        (*first) = newn;
    }

    (*last)->next = (*first);
}
void InsertLast(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no;
    newn->next = NULL;

    if((*first == NULL)  && (*last == NULL))
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        (*last)->next = (newn);
        (*last) = newn;
    }

    (*last)->next = (*first);
}
void DeleteFirst(PPNODE first,PPNODE last)
{   
    PNODE temp = NULL;

    if((*first == NULL)  && (*last == NULL))
    {
        return;
    }
    else if((*first) == (*last))
    {   
        free(*first);

        (*first) = NULL;
        (*last) = NULL;
    }
    else
    {
        temp = (*first);

        (*first) = (*first)->next;

        free(temp);
    }
    (*last)->next = (*first);
}
void DeleteLast(PPNODE first,PPNODE last)
{
    PNODE temp = NULL;

    if((*first == NULL)  && (*last == NULL))
    {
        return;
    }
    else if((*first) == (*last))
    {   
        free(*first);

        (*first) = NULL;
        (*last) = NULL;
    }
    else
    {
        temp = (*first);

        while(temp->next != (*last))
        {
            temp = temp->next;
        }

        free(*last);
        *last = temp;
    }

    (*last)->next = (*first);
}
void Display(PNODE first,PNODE last)
{
    do
    {
        printf("| %d |=>",first->data);

        first = first->next;
    } while ((first) != (last)->next);

    printf("\n");
    
}
int Count(PNODE first,PNODE last)
{   
    int iCount = 0;

    do
    {
        iCount++;

        first = first->next;
    } while ((first) != (last)->next);

    return iCount;
}
void InsertAtPos(PPNODE first,PPNODE last,int no,int pos)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    int iSize = 0;
    int iCnt = 0;

    iSize = Count(*first,*last);

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no;
    newn->next = NULL;

    if(pos < 1 || pos > (iSize + 1))
    {
        printf("Invalid Position :");
        return;
    }

    if(pos == 1)
    {
        InsertFirst(first,last,no);
    }
    else if(pos == (iSize + 1))
    {
        InsertLast(first,last,no);
    }
    else
    {
        temp = (*first);

        for(iCnt = 1;iCnt < pos - 1;iCnt++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;

        temp->next = newn;

        (*last)->next = (*first);
    }

}

void DeleteAtPos(PPNODE first,PPNODE last,int pos)
{
    PNODE temp = NULL;
    PNODE target = NULL;

    int iSize = 0;
    int iCnt = 0;

    iSize = Count(*first,*last);

    if(pos < 1 || pos > (iSize))
    {
        printf("Invalid Position :");
        return;
    }

    if(pos == 1)
    {
        DeleteFirst(first,last);
    }
    else if(pos == (iSize))
    {
        DeleteLast(first,last);
    }
    else
    {
        temp = (*first);

        for(iCnt = 1;iCnt < pos - 1;iCnt++)
        {
            temp = temp->next;
        }

        target = temp->next;

        temp->next = target->next;

        free(target);

        (*last)->next = (*first);
    }
}


int main()
{   
    PNODE head = NULL;
    PNODE tail = NULL;

    int iRet = 0;

    InsertFirst(&head,&tail,51);
    InsertFirst(&head,&tail,41);
    InsertFirst(&head,&tail,31);
    InsertFirst(&head,&tail,21);
    InsertFirst(&head,&tail,11);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of Nodes are : %d\n",iRet);

    InsertLast(&head,&tail,101);
    InsertLast(&head,&tail,111);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of Nodes are : %d\n",iRet);

    DeleteFirst(&head,&tail);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of Nodes are : %d\n",iRet);

    DeleteLast(&head,&tail);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of Nodes are : %d\n",iRet);

    InsertAtPos(&head,&tail,75,3);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of nodes in given list is : %d\n",iRet);

    DeleteAtPos(&head,&tail,3);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of nodes in given list is : %d\n",iRet);


    return 0;
}

File: SinglyCircular_LL/SinglyCL3.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class SinglyCL:
    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    def InsertFirst(self,no):
       
        newn = Node(no)

        if((self.first == None) and (self.last == None)):

            self.first = newn
            self.last = newn

        else:
            newn.next = self.first
            self.first = newn


        self.last.next = self.first
        self.iCount = self.iCount + 1

    def InsertLast(self,no):

        newn = Node(no)

        if((self.first == None) and (self.last == None) ):

            self.first = newn
            self.last = newn

        else:
            
            self.last.next = newn
            self.last = newn


        self.last.next = self.first
        self.iCount = self.iCount + 1

    def DeleteFirst(self):
        
        temp = None

        if((self.first == None) and (self.last == None)):
            return
        
        elif(self.first == self.last):
            
            self.first = None
            self.last = None

            del self.first
            del self.last
        
        else:
            temp = self.first

        self.first = self.first.next

        self.last.next = self.first
        self.iCount = self.iCount - 1

    def DeleteLast(self):
        
        temp = None

        if((self.first == None) and (self.last == None)):
            return
        
        elif(self.first == self.last):
            
            self.first = None
            self.last = None

            del self.first
            del self.last
        
        else:
            
            temp = self.first

            while(temp.next != self.last):
                temp = temp.next

            
            del self.last
            self.last = temp

        self.last.next = self.first
        self.iCount = self.iCount - 1

    def Display(self):

        if self.first is None:
            print("Linked List is empty")
            return
        
        temp = self.first

        while True:
            print("| ", temp.data, " |->", end=" ")
            temp = temp.next

            if temp == self.last.next:
                break

        print()

    def Count(self):
        return self.iCount

    def InsertAtPos(self,no,pos):
        
        newn = None

        newn = Node(no)

        newn.data = no
        newn.next = None

        if(pos < 1 or pos > self.iCount + 1):
            print("Invalid Position")
            return
        
        if(pos == 1):
            self.InsertFirst(no)

        elif(pos == self.iCount + 1):
            self.InsertLast(no)
        
        else:

            temp = self.first

            for i in range(pos - 1):
                temp = temp.next

            newn.next = temp.next

            temp.next = newn

            self.iCount = self.iCount + 1
        

    def DeleteAtPos(self,pos):
    

        if(pos < 1 or pos > self.iCount):
            print("Invalid Position")
            return
        
        if(pos == 1):
            self.DeleteFirst()

        elif(pos == self.iCount):
            self.DeleteLast()
        
        else:

            temp = self.first
            
            for i in range(pos - 1):
                temp = temp.next

            target = temp.next

            temp.next = target.next

            del target

            self.iCount = self.iCount - 1


def main():

    sobj = SinglyCL()

    sobj.InsertFirst(101)
    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.iCount)

    print()

    sobj.InsertLast(101)
    sobj.InsertLast(111)

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.iCount)

    print()

    sobj.DeleteFirst()

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.Count())

    print()

    sobj.DeleteLast()

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.Count())

    print()

    sobj.InsertAtPos(75,4)

    print("Elements of Linked List are :\n")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.Count())

    print()

    sobj.DeleteAtPos(3)

    print("Elements of Linked List are :\n")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.Count())

    print()

if __name__ == "__main__":
    main()